language/move-vm/integration-tests/src/tests/instantiation_tests.rs

```diff
@@ -0,0 +1,846 @@
+        GasStatus::new_unmetered()
+    } else {
+        GasStatus::new(&INITIAL_COST_SCHEDULE, Gas::new(gas_val))
+    }
+}
+
+// With proper gas_val units this function can be used to profile the VM.
+// All code generated is an infinite loop and with 0 unit we run forever which
+// can be used for profiling
+fn main_run(gas_val: u64) {
+    let (res, time) = run_with_module(&mut get_gas_meter(gas_val), load_pop);
+    println!("* load_pop: {}\nStatus: {:?}", time, res);
+    let (res, time) = run_with_module(&mut get_gas_meter(gas_val), vec_pack_simple);
+    println!("* vec_pack_simple: {}\nStatus: {:?}", time, res);
+    let (res, time) = run_with_module(&mut get_gas_meter(gas_val), vec_pack_gen_simple);
+    println!("* vec_pack_gen_simple: {}\nStatus: {:?}", time, res);
+    let (res, time) = run_with_module(&mut get_gas_meter(gas_val), vec_pack_gen_deep);
+    println!("* vec_pack_gen_deep: {}\nStatus: {:?}", time, res);
+    let (res, time) = run_with_module(&mut get_gas_meter(gas_val), simple_gen_exists);
+    println!("* simple_gen_exists: {}\nStatus: {:?}", time, res);
+    let (res, time) = run_with_module(&mut get_gas_meter(gas_val), deep_gen_exists);
+    println!("* deep_gen_exists: {}\nStatus: {:?}", time, res);
+    let (res, time) = run_with_module(&mut get_gas_meter(gas_val), simple_gen_call);
+    println!("* simple_gen_call: {}\nStatus: {:?}", time, res);
+    let (res, time) = run_with_module(&mut get_gas_meter(gas_val), deep_gen_call);
+    println!("* deep_gen_call: {}\nStatus: {:?}", time, res);
+}
+
+#[test]
+fn test_instantiation_no_instantiation() {
+    let (res, ref_time) = run_with_module(&mut get_gas_meter(1000), load_pop);
+    println!("* load_pop: {}\nStatus: {:?}", ref_time, res);
+    assert_eq!(
+        res.err().unwrap().major_status(),
+        StatusCode::OUT_OF_GAS,
+        "Must finish OutOfGas"
+    );
+}
+
+#[test]
+fn test_instantiation_vec_pack_simple() {
+    let (res, ref_time) = run_with_module(&mut get_gas_meter(1000), load_pop);
+    assert_eq!(
+        res.err().unwrap().major_status(),
+        StatusCode::OUT_OF_GAS,
+        "Must finish OutOfGas"
+    );
+    let (res, time) = run_with_module(&mut get_gas_meter(1000), vec_pack_simple);
+    println!("* vec_pack_simple: {}\nStatus: {:?}", time, res);
+    assert_eq!(
+        res.err().unwrap().major_status(),
+        StatusCode::OUT_OF_GAS,
+        "Must finish OutOfGas"
+    );
+    assert!(
+        time < ref_time * 10,
+        "Instantion test taking too long {}",
+        time
+    );
+}
+
+#[test]
+fn test_instantiation_vec_pack_gen_simple() {
+    let (res, ref_time) = run_with_module(&mut get_gas_meter(1000), load_pop);
+    assert_eq!(
+        res.err().unwrap().major_status(),
+        StatusCode::OUT_OF_GAS,
+        "Must finish OutOfGas"
+    );
+    let (res, time) = run_with_module(&mut get_gas_meter(1000), vec_pack_gen_simple);
+    println!("* vec_pack_gen_simple: {}\nStatus: {:?}", time, res);
+    assert_eq!(
```

Comment:
Like this approach! 

Wondering if these tests could become even easier to write with helper function around `run_with_module` that:
- runs the "control" gas check with `load_pop`
- runs the "experiment" gas check with the function provided by the user (e.g., `vec_pack_gen_simple` here
- checks that both control and experiment finish with `OUT_OF_GAS` and the experiment is <10x slower than the control

and of course could parameterize things like the choice of control function, gas budget, 10x threshold.

Reply:
done, I think. Though I am going to add few more things before I make it a regular PR (non draft)